File: src/spectrl/pipeline.py
"""Versioned numeric array encodings for spectrl v3."""

from __future__ import annotations

import re
from collections.abc import Callable
from dataclasses import dataclass

import numpy as np

from ._format import MAX_BLOB_BYTES, MAX_SAFE_INTEGER
from .codecs import quantized, rounded
from .codecs._delta import delta_shuffle, delta_unshuffle
from .codecs.raw import _np_dtype
from .codecs.shuffle import shuffle
from .cv import TYPE_FLOAT32, TYPE_FLOAT64

DESC_ENCODING = 2
DESC_FIDELITY = 7
DESC_PARAMS = 8
DESC_USER_PARAMS = 9
DESC_PROCESSING = 10
DESC_EXTENSIONS = 11

ENCODING_NAMES = {"raw": 0, "byte-shuffle": 1, "modular-delta-shuffle": 2, "quantized": 3, "rounded-float": 4}
_ID = re.compile(r"^[A-Za-z][A-Za-z0-9._-]*:[A-Za-z0-9._/-]+$")


def descriptor(value, names=None):
    """Normalize public operation settings or validate a wire tuple."""
    if isinstance(value, (str, int)) and not isinstance(value, bool):
        value = [value, 1]
    elif isinstance(value, dict):
        if set(value) - {"id", "revision", "parameters"} or "id" not in value:
            raise ValueError("invalid operation descriptor")
        value = [value["id"], value.get("revision", 1), value.get("parameters", {})]
    if not isinstance(value, list) or len(value) not in (2, 3):
        raise ValueError("operation must be [identifier, revision, optional parameters]")
    identifier, revision = value[:2]
    if names and isinstance(identifier, str):
        identifier = names.get(identifier.removeprefix("spectrl:"), identifier)
    if not (
        (type(identifier) is int and 0 <= identifier <= MAX_SAFE_INTEGER)
        or (isinstance(identifier, str) and _ID.fullmatch(identifier))
    ):
        raise ValueError("operation identifier must be a built-in integer or namespaced string")
    if type(revision) is not int or not 1 <= revision <= MAX_SAFE_INTEGER:
        raise ValueError("operation revision must be a positive safe integer")
    params = value[2] if len(value) == 3 else {}
    if not isinstance(params, dict) or not all(isinstance(k, str) for k in params):
        raise ValueError("operation parameters must be a string-keyed map")
    return [identifier, revision, params] if params else [identifier, revision]


@dataclass(frozen=True)
class Encoding:
    """Callbacks encode(array, type, parameters) and decode(bytes, type, count, parameters)."""

    encode: Callable
    decode: Callable
    validate: Callable
    lossless: bool = True
    types: tuple[int, ...] = (1000521, 1000523, 1000519)


ENCODINGS: dict[tuple[int | str, int], Encoding] = {}


def _register(registry, identifier, implementation, revision):
    desc = descriptor([identifier, revision])
    if not isinstance(identifier, str) or not _ID.fullmatch(identifier) or identifier.startswith("spectrl:"):
        raise ValueError("custom registrations require a non-spectrl namespaced identifier")
    key = tuple(desc[:2])
    if key in registry:
        raise ValueError(f"operation is already registered: {identifier}@{revision}")
    registry[key] = implementation


def register_encoding(identifier: str, implementation: Encoding, *, revision: int = 1):
    if not isinstance(implementation, Encoding):
        raise TypeError("implementation must be an Encoding")
    _register(ENCODINGS, identifier, implementation, revision)


def operation(registry, desc):
    desc = descriptor(desc)
    try:
        impl = registry[tuple(desc[:2])]
    except KeyError:
        raise ValueError(f"unsupported operation {desc[0]!r} revision {desc[1]}") from None
    params = desc[2] if len(desc) == 3 else {}
    impl.validate(params)
    return impl, params


def _empty(params):
    if params:
        raise ValueError("this operation does not accept parameters")


def _raw(data, tail, params):
    return np.asarray(data).astype(_np_dtype(tail)).tobytes()


def _from_raw(raw, tail, n, params):
    dtype = np.dtype(_np_dtype(tail))
    if len(raw) != n * dtype.itemsize:
        raise ValueError("decoded byte length does not match dtype and count")
    return np.frombuffer(raw, dtype).copy()


def _transform(forward, inverse):
    return Encoding(
        lambda a, t, p: forward(_raw(a, t, p), np.dtype(_np_dtype(t)).itemsize),
        lambda b, t, n, p: _from_raw(inverse(b, np.dtype(_np_dtype(t)).itemsize), t, n, p),
        _empty,
    )


ENCODINGS.update(
    {
        (0, 1): Encoding(_raw, _from_raw, _empty),
        (1, 1): _transform(shuffle, lambda b, w: shuffle(b, w, inverse=True)),
        (2, 1): _transform(delta_shuffle, delta_unshuffle),
        (3, 1): Encoding(quantized.encode, quantized.decode, quantized.validate, False, (TYPE_FLOAT32, TYPE_FLOAT64)),
        (4, 1): Encoding(rounded.encode, rounded.decode, rounded.validate, False, (TYPE_FLOAT32, TYPE_FLOAT64)),
    }
)


def encode_pipeline(data, tail, enc):
    encoding, params = operation(ENCODINGS, enc)
    if tail not in encoding.types:
        raise ValueError("encoding does not support the declared dtype")
    blob = encoding.encode(data, tail, params)
    if not isinstance(blob, bytes) or len(blob) > min(MAX_BLOB_BYTES, 64 + 16 * len(data)):
        raise ValueError("encoding exceeds the intermediate byte limit")
    return blob, 0 if encoding.lossless else 1


def decode_pipeline(blob, tail, count, enc, fidelity):
    encoding, params = operation(ENCODINGS, enc)
    if tail not in encoding.types or fidelity != (0 if encoding.lossless else 1):
        raise ValueError("encoding dtype or fidelity declaration mismatch")
    if len(blob) > min(MAX_BLOB_BYTES, 64 + 16 * count):
        raise ValueError("encoded data exceeds byte limit")
    result = encoding.decode(blob, tail, count, params)
    if not isinstance(result, np.ndarray) or result.ndim != 1 or len(result) != count:
        raise ValueError("encoding returned an invalid array shape")
    if result.dtype != np.dtype(_np_dtype(tail)):
        raise ValueError("encoding returned an incorrect dtype")
    return result

File: js/src/pipeline.ts
/** Versioned numeric array encodings for v3. */
import { byteShuffle, byteUnshuffle, decodeRaw, encodeRaw, type NumArray } from "./codecs.js"
import { deltaShuffle, deltaUnshuffle } from "./delta.js"
import { MAX_BLOB_BYTES } from "./format.js"
import { encodeQuantized, decodeQuantized, validateQuantized } from "./quantized.js"
import { encodeRounded, decodeRounded, validateRounded } from "./rounded.js"

export type Parameters = Record<string, unknown>
export type Operation = [number | string, number, Parameters?]
export type OperationOption = number | string | Operation | { id: number | string, revision?: number, parameters?: Parameters }
export interface Encoding {
  encode: (data: NumArray, type: number, params: Parameters) => Uint8Array
  decode: (bytes: Uint8Array, type: number, count: number, params: Parameters) => NumArray
  validate: (params: Parameters) => void
  lossless: boolean
  types: readonly number[]
}
export const encodingNames: Record<string, number> = { raw: 0, "byte-shuffle": 1, "modular-delta-shuffle": 2, quantized: 3, "rounded-float": 4 }
const namespace = /^[A-Za-z][A-Za-z0-9._-]*:[A-Za-z0-9._/-]+$/
export function descriptor(value: OperationOption, names?: Record<string, number>): Operation {
  let tuple: unknown[]
  if (typeof value === "number" || typeof value === "string") tuple = [value, 1]
  else if (Array.isArray(value)) tuple = value
  else {
    if (!value || typeof value !== "object" || Object.keys(value).some(k => !["id", "revision", "parameters"].includes(k))) throw Error("invalid operation descriptor")
    tuple = [value.id, value.revision ?? 1, value.parameters ?? {}]
  }
  if (![2, 3].includes(tuple.length)) throw Error("operation requires identifier, revision, and optional parameters")
  let [id, revision, params = {}] = tuple
  if (names && typeof id === "string") id = names[id.replace(/^spectrl:/, "")] ?? id
  if (!(typeof id === "number" && Number.isSafeInteger(id) && id >= 0) && !(typeof id === "string" && namespace.test(id))) throw Error("invalid operation identifier")
  if (typeof revision !== "number" || !Number.isSafeInteger(revision) || revision < 1) throw Error("invalid operation revision")
  if (params instanceof Map) {
    if ([...params.keys()].some(k => typeof k !== "string")) throw Error("operation parameters require string keys")
    params = Object.fromEntries(params)
  }
  if (!params || typeof params !== "object" || Array.isArray(params) || params instanceof Uint8Array) throw Error("operation parameters must be a map")
  return Object.keys(params).length ? [id as number | string, revision, params as Parameters] : [id as number | string, revision]
}
export const encodings = new Map<string, Encoding>()
export const operationKey = (op: Operation) => JSON.stringify(op.slice(0, 2))
function register<T>(registry: Map<string, T>, id: string, impl: T, revision: number) {
  const desc = descriptor([id, revision])
  if (!namespace.test(id) || id.startsWith("spectrl:")) throw Error("custom registration requires a non-spectrl namespace")
  const key = operationKey(desc)
  if (registry.has(key)) throw Error("operation already registered")
  registry.set(key, impl)
}
export const registerEncoding = (id: string, impl: Encoding, revision = 1) => register(encodings, id, impl, revision)
export function operation<T extends { validate: (p: Parameters) => void }>(registry: Map<string, T>, desc: Operation): [T, Parameters] {
  const impl = registry.get(operationKey(desc))
  if (!impl) throw Error(`unsupported operation ${desc[0]} revision ${desc[1]}`)
  const params = desc[2] ?? {}
  impl.validate(params)
  return [impl, params]
}
const empty = (p: Parameters) => { if (Object.keys(p).length) throw Error("operation accepts no parameters") }
const width = (type: number) => type === 1000523 ? 8 : 4
for (const id of [0, 1, 2]) encodings.set(operationKey([id, 1]), {
  lossless: true, types: [1000521, 1000523, 1000519], validate: empty,
  encode: (a, t) => { const raw = encodeRaw(a, t)
    return id === 0 ? raw : id === 1 ? byteShuffle(raw, width(t)) : deltaShuffle(raw, width(t))
  },
  decode: (b, t, n) => decodeRaw(id === 0 ? b : id === 1 ? byteUnshuffle(b, width(t)) : deltaUnshuffle(b, width(t)), t),
})
encodings.set(operationKey([3, 1]), { encode: encodeQuantized, decode: decodeQuantized, validate: validateQuantized, lossless: false, types: [1000523] })
encodings.set(operationKey([4, 1]), { encode: encodeRounded, decode: decodeRounded, validate: validateRounded, lossless: false, types: [1000521, 1000523] })
export function encodePipeline(data: NumArray, type: number, enc: Operation) {
  const [e, params] = operation(encodings, enc)
  if (!e.types.includes(type)) throw Error("encoding does not support dtype")
  const blob = e.encode(data, type, params)
  if (!(blob instanceof Uint8Array) || blob.length > Math.min(MAX_BLOB_BYTES, 64 + 16 * data.length)) throw Error("encoding exceeds intermediate limit")
  return { blob, fidelity: e.lossless ? 0 : 1 }
}
export function decodePipeline(blob: Uint8Array, type: number, count: number, enc: Operation, fidelity: number): NumArray {
  const [e, params] = operation(encodings, enc)
  if (!e.types.includes(type) || fidelity !== (e.lossless ? 0 : 1)) throw Error("encoding dtype or fidelity mismatch")
  if (blob.length > Math.min(MAX_BLOB_BYTES, 64 + 16 * count)) throw Error("encoded byte limit exceeded")
  const out = e.decode(blob, type, count, params)
  if (out.length !== count || !(type === 1000523 ? out instanceof Float64Array : type === 1000521 ? out instanceof Float32Array : out instanceof Int32Array)) throw Error("decoded shape or dtype mismatch")
  return out
}
